scalafmt-core/shared/src/main/scala/org/scalafmt/util/MarkdownFile.scala

```diff
@@ -0,0 +1,119 @@
+
+    private def substringWithAdaptedEnd(start: Int, end: Int): String = {
+      val adaptedEnd = math.max(start, end)
+      text.substring(start, adaptedEnd)
+    }
+
+    private def newCodeFence(
+        state: State.CodeFence,
+        backtickStart: Int,
+        backtickEnd: Int
+    ): MarkdownPart.CodeFence = {
+      val open = substringWithAdaptedEnd(
+        state.start,
+        state.start + state.backticks.length()
+      )
+      val lastIndexOfOpeningBackTicks = state.start + state.backticks.length()
+      val info = substringWithAdaptedEnd(
+        lastIndexOfOpeningBackTicks,
+        lastIndexOfOpeningBackTicks + state.info.length()
+      )
+      val adaptedBacktickStart = math.max(0, backtickStart - 1)
+      val body = substringWithAdaptedEnd(
+        lastIndexOfOpeningBackTicks + info.length(),
+        adaptedBacktickStart
+      )
+      val close = substringWithAdaptedEnd(adaptedBacktickStart, backtickEnd)
+      MarkdownPart.CodeFence(open, info, body, close)
+    }
+    def acceptParts(): List[MarkdownPart] = {
+      var state: State = State.Text
+      val parts = mutable.ListBuffer.empty[MarkdownPart]
+      var curr = 0
+      text.linesWithSeparators
+        .foreach { line =>
+          val end = curr + line.length()
+          state match {
+            case State.Text =>
+              if (line.startsWith("```")) {
+                val backticks = line.takeWhile(_ == '`')
+                val info = line.substring(backticks.length())
+                state = State.CodeFence(curr, backticks, info)
+              } else {
+                parts += MarkdownPart.Text(substringWithAdaptedEnd(curr, end))
+              }
+            case s: State.CodeFence =>
+              if (
+                line.startsWith(s.backticks) &&
+                line.forall(ch => ch == '`' || ch.isWhitespace)
+              ) {
+                parts += newCodeFence(s, curr, end)
+                state = State.Text
+              }
+          }
+          curr = end
+        }
+      state match {
+        case s: State.CodeFence =>
+          parts += newCodeFence(s, text.length(), text.length())
+        case _ =>
+      }
+      parts.toList
+    }
+  }
+
+  def parse(input: Input): MarkdownFile =
+    MarkdownFile(new Parser(input).acceptParts())
+
+}
+
+sealed abstract class MarkdownPart {
+  final def renderToString(out: StringBuilder): Unit =
+    this match {
+      case MarkdownPart.Text(value) =>
+        out.append(value)
+      case fence: MarkdownPart.CodeFence =>
+        out.append(fence.openBackticks)
+        out.append(fence.info)
+        fence.newBody match {
+          case None =>
+            out.append(fence.body)
+          case Some(newBody) =>
+            out.append(newBody)
+        }
+        out.append(fence.closeBackticks)
+    }
+}
+object MarkdownPart {
+  final case class Text(value: String) extends MarkdownPart
+  final case class CodeFence(
+      openBackticks: String,
```

Comment:
are open backticks ever going to be different from close?

Reply:
It sounds like that's against spec, so not defined behavior.
https://meta.stackoverflow.com/a/409129/1128869
We could do a few different things

1. Ignore the discrepancy, and just copy the front backticks onto the end
2. Preserve the unbalanced backticks
3. Error out if unbalanced

Currently we're doing # 2. I don't have any strong feelings about this.

Reply:
Added a test that codifies the current behavior.
Will update it if a different choice is made.